synapse/rest/client/v1/room.py

```diff
@@ -267,6 +284,108 @@ def on_PUT(self, request, room_id, event_type, txn_id):
         )
 
 
+class RoomBulkSendEventRestServlet(TransactionRestServlet):
+    def __init__(self, hs):
+        super().__init__(hs)
+        self.event_creation_handler = hs.get_event_creation_handler()
+        self.auth = hs.get_auth()
+
+        self._msc2716_enabled = hs.config.experimental.msc2716_enabled
+
+    def register(self, http_server):
+        # /rooms/$roomid/bulksend
+        PATTERNS = "/rooms/(?P<room_id>[^/]*)/bulksend"
+        register_txn_path(self, PATTERNS, http_server, with_get=True)
+
+    async def on_POST(self, request, room_id):
```

Comment:
Bulk send endpoint as described in https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r598896168

```json
{
    "events": [ ... ],
    "state_events_at_start": [ ... ]
}
```

Currently paired with the Complement tests, see `Backfilled historical events resolve with proper state` (and `backfillBulkHistoricalMessagesAtTime`), the backfilled messages will send without a 403 but they don't appear in the DAG.

Even the first historical message does not appear in the DAG which should be the same as we were doing before. I can see it grabbing the successor event and setting the depth correctly.

Reply:
I'd like to use the `self.event_creation_handler.create_and_send_nonmember_event` normal flow but the `self.auth.check_from_context` down the line is not picking up the outlier state and denying because it claims the user is not in the room.

Maybe need to make sure the `context` has the outlier referenced in it 🤔

Reply:
Got the events showing up in the DAG! Just skipped all of the auth and did `persist_event`. Still want to go back and see if I can use the normal `create_and_send_nonmember_event` flow and setup the context correctly so the outlier state auths it correctly.

Reply:
> Got the events showing up in the DAG! Just skipped all of the auth and did `persist_event`. Still want to go back and see if I can use the normal `create_and_send_nonmember_event` flow and setup the context correctly so the outlier state auths it correctly.

Yeah, that would probably be good. We might want to factor some things out as I imagine a bunch of the logic and checks just don't make any sense for outlier state, but that's all just refactoring

Reply:
@erikjohnston I got everything working with what I think are authed events. How does this sort of thing look? I have a few areas to clean up but want to get a general sense if this works?

Testing against the complement test `Backfilled historical events resolve with proper state`, https://github.com/matrix-org/complement/pull/68/files

Reply:
@erikjohnston Cleaned up the PR and is working well with the [Complement tests](https://github.com/matrix-org/complement/pull/68) to insert multiple chunks of historical messages! 😀 

<img width="460" src="https://user-images.githubusercontent.com/558581/114769458-dc4d2880-9d2f-11eb-9a50-e2181b88916c.png">